core/node/state_keeper/src/updates/mod.rs

```diff
@@ -77,6 +79,27 @@ impl UpdatesManager {
         self.base_system_contract_hashes
     }
 
+    pub(crate) fn get_next_l2_block_timestamp_as_mut(&mut self) -> &mut u64 {
+        if let Some(l2_block_params) = self.next_l2_block_params.as_mut() {
+            return &mut l2_block_params.timestamp;
+        }
+        &mut self.l2_block.timestamp
```

Comment:
AFAIU, this fallback is logically incorrect, but it's never used. It's incorrect because changing a timestamp of the 1st block in a batch must be done together with changing the timestamp of the batch (they should always be equal), and neither can really change because a VM instance for the batch is already initialized at this point. It seems that there's a neat opportunity to combine this method with `has_next_block_params()`:

```rust
pub(crate) fn next_l2_block_timestamp_mut(&mut self) -> Option<&mut u64> {
    self.next_l2_block_params.as_mut().map(|params| &mut params.timestamp)
}

// usage in state keeper logic like `process_l1_batch()`
if let Some(timestamp) = updates_manager.next_l2_block_timestamp_mut() {
    self.io.update_next_l2_block_timestamp(timestamp);
}
```

Reply:
Make totally sense! Fixed